From b5d3510eb058f5fd5f24f7ade1490e2592844d7b Mon Sep 17 00:00:00 2001
From: Xu Qiang <xuqiang36@huawei.com>
Date: Mon, 23 May 2022 21:18:28 +0800
Subject: [PATCH] irq-gic-v3-its: It can't be initialized when the GICR had
 been cut
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: b5d3510eb058f5fd5f24f7ade1490e2592844d7b
Modified-by-SEL: No


ascend inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I554T5
CVE: NA

------------

In FPGA, We need to check if the gicr has been cut,
and if it is, it can't be initialized

Signed-off-by: Xu Qiang <xuqiang36@huawei.com>
Reviewed-by: Liao Chang <liaochang1@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/irqchip/irq-gic-v3-its.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/drivers/irqchip/irq-gic-v3-its.c b/drivers/irqchip/irq-gic-v3-its.c
index e04e0166eac3..54662b36dd6a 100644
--- a/drivers/irqchip/irq-gic-v3-its.c
+++ b/drivers/irqchip/irq-gic-v3-its.c
@@ -3390,6 +3390,7 @@ static void its_cpu_init_lpis_others(void __iomem *rbase, int cpu)
 static void its_cpu_init_collection_others(void __iomem *rbase,
 					   phys_addr_t phys_base, int cpu)
 {
+	u32 count;
 	struct its_node *its;
 
 	if (!init_all_gicr)
@@ -3418,6 +3419,32 @@ static void its_cpu_init_collection_others(void __iomem *rbase,
 			target = GICR_TYPER_CPU_NUMBER(target) << 16;
 		}
 
+		dsb(sy);
+
+		/* In FPGA, We need to check if the gicr has been cut,
+		 * and if it is, it can't be initialized
+		 */
+		count = 2000;
+		while (1) {
+			if (readl_relaxed(rbase + GICR_SYNCR) == 0)
+				break;
+
+			count--;
+			if (!count) {
+				pr_err("this gicr does not exist, or it's abnormal:%pK\n",
+					&phys_base);
+				break;
+			}
+			cpu_relax();
+			udelay(1);
+		}
+
+		if (count == 0)
+			break;
+
+		pr_info("its init other collection table, ITS:%pK, GICR:%pK, coreId:%u\n",
+			&its->phys_base, &phys_base, cpu);
+
 		/* Perform collection mapping */
 		its->collections[cpu].target_address = target;
 		its->collections[cpu].col_id = cpu;
-- 
2.33.0

